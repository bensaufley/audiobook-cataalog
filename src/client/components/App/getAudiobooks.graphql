query getAudiobooks($startId: ID) {
  getAudiobooks(start: $startId) {
    id
    name
    authors {
      id
      author {
        lastName
        firstName
      }
      meta
    }
  }
}
