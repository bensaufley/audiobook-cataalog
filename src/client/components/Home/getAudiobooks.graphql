query getAudiobooks($startId: ID) {
  getAudiobooks(start: $startId) {
    ...audiobook
  }
}

fragment audiobook on Audiobook {
  id
  duration
  filepath
  name
  year
  genres {
    name
  }
  authors {
    id
    author {
      lastName
      firstName
    }
    meta
  }
}
