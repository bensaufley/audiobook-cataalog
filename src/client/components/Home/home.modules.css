.h1 {
  margin-left: auto;
  margin-right: auto;
  max-width: 75rem;
  padding: 0 1rem;
}

.audiobook-container {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 0.675rem;

  @media screen and (min-width: 32em) {
    grid-template-columns: repeat(3, 1fr);
  }

  @media screen and (min-width: 48rem) {
    grid-template-columns: repeat(4, 1fr);
  }

  @media screen and (min-width: 64rem) {
    grid-template-columns: repeat(5, 1fr);
  }

  @media screen and (min-width: 80rem) {
    grid-template-columns: repeat(6, 1fr);
  }

  @media screen and (min-width: 96rem) {
    grid-template-columns: repeat(7, 1fr);
  }

  @media screen and (min-width: 112rem) {
    grid-template-columns: repeat(8, 1fr);
  }
}
