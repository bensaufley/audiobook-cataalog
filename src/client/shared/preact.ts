import type { Ref } from 'preact';

// eslint-disable-next-line import/prefer-default-export
export const mergeRefs = <T extends HTMLElement>(...refs: (Ref<T> | null | undefined)[]): Ref<T> | null => {
  const filteredRefs = refs.filter((ref) => !!ref);
  if (!filteredRefs.length) return null;
  if (filteredRefs.length === 1) return filteredRefs[0]!;

  return (el: T | null) => {
    filteredRefs.forEach((ref) => {
      if (typeof ref === 'function') ref(el);
      // eslint-disable-next-line no-param-reassign
      else ref.current = el;
    });
  };
};
