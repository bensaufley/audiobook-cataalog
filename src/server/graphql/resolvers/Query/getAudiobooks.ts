import getCollection from '~server/components/db/getCollection';
import type { QueryResolvers } from '~server/graphql/resolvers/types';

const getAudiobooks: QueryResolvers['getAudiobooks'] = async (_, _args, { db }) => {
  const audiobooks = await getCollection(db, 'audiobooks');
  return audiobooks.find().sort({ name: 1 }).toArray();
};

export default getAudiobooks;
