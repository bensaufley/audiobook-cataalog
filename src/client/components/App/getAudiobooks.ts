// Generated file. Do not edit!

import * as Urql from '@urql/preact';
import gql from 'graphql-tag';

import type * as SchemaTypes from '~graphql/schema';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type GetAudiobooksQueryVariables = SchemaTypes.Exact<{
  startId: SchemaTypes.Maybe<SchemaTypes.Scalars['ID']>;
}>;

export type GetAudiobooksQuery = {
  __typename?: 'Query';
  getAudiobooks: Array<{
    __typename?: 'Audiobook';
    id: string;
    name: string;
    authors: Array<{
      __typename?: 'AudiobookAuthor';
      id: string;
      meta: SchemaTypes.Maybe<string>;
      author: { __typename?: 'Author'; lastName: string; firstName: SchemaTypes.Maybe<string> };
    }>;
  }>;
};

export const GetAudiobooksDocument = gql`
  query getAudiobooks($startId: ID) {
    getAudiobooks(start: $startId) {
      id
      name
      authors {
        id
        author {
          lastName
          firstName
        }
        meta
      }
    }
  }
`;

export function useGetAudiobooksQuery(
  options: Omit<Urql.UseQueryArgs<GetAudiobooksQueryVariables>, 'query'> = {}
) {
  return Urql.useQuery<GetAudiobooksQuery>({ query: GetAudiobooksDocument, ...options });
}
