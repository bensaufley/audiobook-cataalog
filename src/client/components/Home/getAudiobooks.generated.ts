// Generated file. Do not edit!

import * as Urql from '@urql/preact';
import gql from 'graphql-tag';

import type * as SchemaTypes from '~graphql/schema.generated';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type GetAudiobooksQueryVariables = SchemaTypes.Exact<{
  startId: SchemaTypes.Maybe<SchemaTypes.Scalars['ID']>;
}>;

export type GetAudiobooksQuery = {
  __typename?: 'Query';
  getAudiobooks: Array<{ __typename?: 'Audiobook' } & AudiobookFragment>;
};

export type AudiobookFragment = {
  __typename?: 'Audiobook';
  id: string;
  duration: number;
  filename: string;
  name: string;
  year: SchemaTypes.Maybe<number>;
  genres: Array<{ __typename?: 'Genre'; name: string }>;
  authors: Array<{
    __typename?: 'AudiobookAuthor';
    id: string;
    meta: SchemaTypes.Maybe<string>;
    author: { __typename?: 'Author'; lastName: string; firstName: SchemaTypes.Maybe<string> };
  }>;
};

export const AudiobookFragmentDoc = gql`
  fragment audiobook on Audiobook {
    id
    duration
    filename
    name
    year
    genres {
      name
    }
    authors {
      id
      author {
        lastName
        firstName
      }
      meta
    }
  }
`;
export const GetAudiobooksDocument = gql`
  query getAudiobooks($startId: ID) {
    getAudiobooks(start: $startId) {
      ...audiobook
    }
  }
  ${AudiobookFragmentDoc}
`;

export function useGetAudiobooksQuery(
  options: Omit<Urql.UseQueryArgs<GetAudiobooksQueryVariables>, 'query'> = {},
) {
  return Urql.useQuery<GetAudiobooksQuery>({ query: GetAudiobooksDocument, ...options });
}
