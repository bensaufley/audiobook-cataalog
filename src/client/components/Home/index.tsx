import { FunctionComponent, h } from 'preact';

import { useGetAudiobooksQuery } from '~client/components/App/getAudiobooks';

const Home: FunctionComponent = () => {
  const [
    { data: { getAudiobooks: audioBooks } = { getAudiobooks: [] }, error },
  ] = useGetAudiobooksQuery();

  return (
    <div>
      <h1>Audiobook Catalog</h1>
      <pre>{JSON.stringify(error ? error.toString() : audioBooks, null, 2)}</pre>
    </div>
  );
};

export default Home;
