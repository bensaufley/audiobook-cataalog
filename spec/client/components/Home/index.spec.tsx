import { ObjectIDMock } from 'graphql-scalars';
import { h } from 'preact';
import { fromValue } from 'wonka';

import Home from '~client/components/Home';
import { GetAudiobooksQuery } from '~client/components/Home/getAudiobooks';
import { mockClient, renderWithProviders } from '~spec/support/spec-helpers';

describe('~client/components/Home', () => {
  it('renders without exploding', () => {
    const stubbedQueryResult: GetAudiobooksQuery = {
      getAudiobooks: [
        {
          id: ObjectIDMock(),
          cover: 'data:image/jpeg;base64,faoiweurpoaewurpoiaewupornawepiofweiohnoipewaf',
          name: 'Ghost in the Wires',
          year: 2011,
          genres: [],
          authors: [
            {
              id: ObjectIDMock(),
              author: {
                lastName: 'Mitnick',
                firstName: 'Kevin',
              },
              meta: null,
            },
            {
              id: ObjectIDMock(),
              author: {
                lastName: 'Simon',
                firstName: 'William L.',
              },
              meta: null,
            },
          ],
        },
        {
          id: ObjectIDMock(),
          cover: null,
          name: 'The Echo Wife',
          year: 2021,
          genres: [],
          authors: [
            {
              id: ObjectIDMock(),
              author: {
                lastName: 'Gailey',
                firstName: 'Sarah',
              },
              meta: null,
            },
          ],
        },
        {
          id: ObjectIDMock(),
          cover: null,
          name: 'The Theory of Everything: The Quest to Explain All Reality',
          year: 2018,
          genres: [],
          authors: [
            {
              id: ObjectIDMock(),
              author: {
                lastName: 'Lincoln',
                firstName: 'Don',
              },
              meta: null,
            },
            {
              id: ObjectIDMock(),
              author: {
                lastName: 'Courses',
                firstName: 'The Great',
              },
              meta: null,
            },
          ],
        },
        {
          id: ObjectIDMock(),
          cover: null,
          name: 'Vagabonds',
          year: 2020,
          genres: [],
          authors: [
            {
              id: ObjectIDMock(),
              author: {
                lastName: 'Jingfang',
                firstName: 'Hao',
              },
              meta: null,
            },
            {
              id: ObjectIDMock(),
              author: {
                lastName: 'Liu',
                firstName: 'Ken',
              },
              meta: 'translator',
            },
          ],
        },
      ],
    };
    const { container } = renderWithProviders({
      client: mockClient({
        executeQuery: jest.fn(() =>
          fromValue({
            data: stubbedQueryResult,
          }),
        ) as any,
      }),
    })(<Home />);

    expect(container).toMatchSnapshot();
  });
});
