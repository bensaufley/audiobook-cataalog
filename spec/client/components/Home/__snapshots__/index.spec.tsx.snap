// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`~client/components/Home renders without exploding 1`] = `
<div>
  <div>
    <h1>
      Audiobook Catalog
    </h1>
    <div>
      <img
        alt="Ghost in the Wires"
        src="data:image/jpeg;base64,faoiweurpoaewurpoiaewupornawepiofweiohnoipewaf"
      />
      <p>
        Ghost in the Wires
      </p>
      <p>
        by
         
        Kevin
         
        Mitnick
        , 
        William L.
         
        Simon
      </p>
      <p>
        2011
      </p>
    </div>
    <div>
      <p>
        The Echo Wife
      </p>
      <p>
        by
         
        Sarah
         
        Gailey
      </p>
      <p>
        2021
      </p>
    </div>
    <div>
      <p>
        The Theory of Everything: The Quest to Explain All Reality
      </p>
      <p>
        by
         
        Don
         
        Lincoln
        , 
        The Great
         
        Courses
      </p>
      <p>
        2018
      </p>
    </div>
    <div>
      <p>
        Vagabonds
      </p>
      <p>
        by
         
        Hao
         
        Jingfang
        , 
        Ken
         
        Liu
         (translator)
      </p>
      <p>
        2020
      </p>
    </div>
  </div>
</div>
`;
